### ISSUE TITLE
[v4] Pass [input,out] to onCreate/constructor lifecycle method

### ISSUE THREAD

**Issue by mindeavor** [marko-js/marko#568]
## New Feature

Following up on [this thread](https://github.com/marko-js/marko/issues/546#issuecomment-274704445), it would be nice to have a lifecycle method that runs once and receives `input` in order to construct the component's initial state.

### Context

Currently the way to set initial state is this:

```jsx
class {
  onInput(input) {
    this.state = { count: input.initialCount }
  }
  inc(){ this.state.count++ }
}
<div>
  Current count: ${ state.count }
  <button on-click('inc')>Increment</button>
</div>
```

But the problem is `onInput` runs more than once – once when the component is constructed and rendered, but also when the component's **parent** is rendered. In the above example, the latter will cause the component to reset its state.

### Possible Implementation

Pass `input` to `onCreate` or `constructor` ? Some lifecycle method that runs once and only once.
